File: Problem_1 - 1.py
"""
Coin Change - 1 

Time Complexity : O(2^n) where n is number of denominations. 
Space Complexity : O(m+n).
    
"""

class Solution:
    def helper(self, coins, amount, index, min_coin):
        
        if amount == 0: return min_coin
        if amount < 0 or index == len(coins): return -1 
        
        case1 = self.helper(coins, amount - coins[index], index, min_coin + 1)
        case2 = self.helper(coins, amount, index + 1, min_coin)
        
        if case1 == - 1: return case2 
        if case2 == -1: return case1 
        
        return min(case1, case2)
        
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0 
        if not coins or len(coins) == 0:
            return 0 
        
        return self.helper(coins, amount, 0, 0)

File: problem11.java
class totalCoinChanges{

    public int coinChange(int[] nums, int amount){
        if(nums == null || nums.length == 0) return 0;

        return helper(nums, amount, 0, 0);
    }

    private int helper(int[] nums, int amount, int index, int total){
        // base
        if(amount == 0){
            return total;
        }
        if(amount < 0 || index == nums.length){
            return -1;
        }

        //logic
        int case1 = helper(nums, amount, index+1, total);

        int case2 = helper(nums, amount- nums[index], index, total+1);

        if(case1 == -1){
            return case2;
        }
        if(case2 == -1){
            return case1;
        }
        return Math.min(case1, case2);
    }

    public static void main(String[] args) {
        totalCoinChanges tcc = new totalCoinChanges();
        int[] nums = {1,2,5};
        int amount = 11;
        System.out.println(tcc.coinChange(nums, amount));
    }
}